### PR TITLE
Enhance bodi exceptions to include name of interface when duplicate registrations attempted

### PR DESCRIPTION
**Pull request by clrudolphi**
Once an interface has been registered and an instance of an object for that registration has been resolved, BoDi does not allow that registration to be altered. The exception message however is vague and does not indicate which Interface registration attempt caused the exception. 

### 🤔 What's changed?

The error messages given to the BoDi exceptions have been enhanced to include type/interface information.

### ⚡️ What's your motivation? 

Ease of diagnosing error situations when enhancing Reqnroll and Plugins.

### 🏷️ What kind of change is this?

- :bank: Refactoring/debt/DX (improvement to code design, tooling, etc. without changing behaviour)

### 📋 Checklist:

- [X ] I've changed the behaviour of the code
  - [X ] I have added/updated tests to cover my changes.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.
- [ ] Users should know about my change
  - [ X] I have added an entry to the "[vNext]" section of the [**CHANGELOG**](../blob/main/CHANGELOG.md), linking to this pull request & included my GitHub handle to the release contributors list.

### CODE REVIEW

File: Reqnroll/BoDi/ObjectContainer.cs [hunk 1]
```diff
@@ -500,7 +500,7 @@ public bool IsRegistered(Type type, string name = null)
     private void AssertNotResolved(RegistrationKey interfaceType)
     {
         if (_resolvedKeys.Contains(interfaceType))
-            throw new ObjectContainerException("An object has been resolved for this interface already.", null);
+            throw new ObjectContainerException($"An object has been resolved for this interface ({interfaceType.ToString()}) already.", null);
```
Comment (Code-Grump) [inline Reqnroll/BoDi/ObjectContainer.cs:503]: The grammar here is a little off. Here's an alternative suggestion:

```suggestion
            throw new ObjectContainerException($"An object has been resolved for the interface \"{interfaceType.ToString()}\" already.", null);
```

File: Tests/Reqnroll.RuntimeTests/BoDi/RegisterTypeTests.cs [hunk 2]
```diff
@@ -65,7 +65,8 @@ public void ShouldNotAllowOverrideRegistrationAfterResolve()
 
             // when 
             Action act = () => container.RegisterTypeAs<SimpleClassWithDefaultCtor, IInterface1>();
-            act.Should().ThrowExactly<ObjectContainerException>();
+            var e = act.Should().ThrowExactly<ObjectContainerException>();
+            e.Which.Message.Should().Contain("IInterface1");
```
Comment (Code-Grump) [inline Tests/Reqnroll.RuntimeTests/BoDi/RegisterTypeTests.cs:69]: There's a `.And` extension that lets you chain these assertions:

```suggestion
            act.Should().ThrowExactly<ObjectContainerException>()
                .And.Message.Should().Contain("IInterface1");
```

**Review by Code-Grump — COMMENTED** [review]
A couple of small suggestions to consider; otherwise looks good.

**Comment by clrudolphi** [comment]
> A couple of small suggestions to consider; otherwise looks good.

@Code-Grump , thanks for the suggestions!

File: Reqnroll/BoDi/ObjectContainer.cs [hunk 3]
```diff
@@ -364,7 +364,7 @@ public IStrategyRegistration RegisterTypeAs<TType, TInterface>(string name = nul
     public IStrategyRegistration RegisterTypeAs(Type implementationType, Type interfaceType, string name = null)
     {
         if (!IsValidTypeMapping(implementationType, interfaceType))
-            throw new InvalidOperationException("type mapping is not valid");
+            throw new InvalidOperationException($"type mapping is not valid: {implementationType} -> {interfaceType}");
```
Comment (ajeckmans) [inline Reqnroll/BoDi/ObjectContainer.cs:367]: This error message is quite technical. Consider rewording it to something like `The provided type mapping is not valid. 
The {implementationType} is not assignable to {interfaceType}`

Comment (clrudolphi) [reply Reqnroll/BoDi/ObjectContainer.cs]: Agreed. Modified.

**Review by gasparnagy — APPROVED** [review]
thx!

**Comment by gasparnagy** [comment]
Released in 2.3.0: https://github.com/reqnroll/Reqnroll/releases/tag/v2.3.0

**Pull request merged**